node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -319,11 +320,16 @@ impl<C: Context> ActiveValidator<C> {
         vec![Effect::ScheduleTimer(self.next_timer)]
     }
 
-    /// Returns the earliest timestamp where we can cast our next vote without equivocating, i.e.
-    /// the timestamp of our previous vote, or 0 if there is none.
+    /// Returns the earliest timestamp where we can cast our next vote: It can't be earlier than
+    /// our previous vote, and it can't be the third vote in a single round.
     fn earliest_vote_time(&self, state: &State<C>) -> Timestamp {
         self.latest_vote(state)
-            .map_or_else(Timestamp::zero, |vh| vh.timestamp)
+            .map_or_else(Timestamp::zero, |vote| {
+                vote.previous().map_or(vote.timestamp, |vh2| {
+                    let vote2 = state.vote(vh2);
+                    vote.timestamp.max(vote2.round_id() + vote2.round_len())
```

Comment:
Hmm, so if an incoming vote has `timestamp` earlier than our last vote we will confirm it with our next vote?

Another question: the "earliest vote time" calculate here uses our current round length (exp) – what if we decide to change the round exponent? For example, make it longer, won't we schedule the timer for the wrong "moment"?

Reply:
Yes, adapting round exponents is not yet implemented, and will need some changes to this code.